main.go

```diff
@@ -129,19 +129,29 @@ func reportFile(out io.Writer, f string) {
 
 func walk(p, ext string, exclude []string, dry bool, out io.Writer) (int, error) {
 	var code int
-	return code, filepath.Walk(p, func(path string, info os.FileInfo, err error) error {
-		if err != nil {
+	var err error
+	if e := filepath.Walk(p, func(path string, info os.FileInfo, walkErr error) error {
+		if walkErr != nil {
 			code = exitFailedToWalkPath
-			return err
+			return walkErr
 		}
 
 		if info.IsDir() && stringInSlice(info.Name(), exclude) {
 			return filepath.SkipDir
 		}
+		c, e := addOrCheckLicense(path, ext, info, dry, out)
+		if c > 0 {
+			code = c
+		}
+		if err != nil {
```

Comment:
IIUC this could lead to a case where the returned `error` value is unrelated to the error code. Is that something you want?

I think it would be easier to reason about if there were only a single value indicating that an error occurred, but it's a slightly bigger change. You could have an `error` type that has an `ExitCode()` method. For example:

```
type exitCoder interface {
    ExitCode() int
}

type exitError struct {
    exitCode int
    cause error
}

func (e exitError) Error() string { ... }
func (e exitError) ExitCode() int { return e.exitCode }

// exitStatus is determines what status to set when the process exits.
func exitStatus(err error) int {
  // if err == nil return 0
  // if err implements exitCoder return exitCode()
  // else return non-zero
}

func main() {
  ...
  os.Exit(exitStatus(err))
}
```

Reply:
I had already thought about it but didn't bother going about that route since it meant making it a bigger change, but it might make sense to do that.